Sources/Basic/Path.swift

```diff
@@ -0,0 +1,473 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import libc
+import POSIX
+
+
+/// Path separator character (always `/`).  If a path separator occurs at the
+/// beginning of a path, the path is considered to be absolute.
+public let PathSeparatorCharacter: Character = "/"
+
+/// User directory replacement character (always `~`).  When it occurs at the
+/// beginning of the first component of a path (optionally followed by a user
+/// name), that first path component is replaced by the absolute path of the
+/// home directory of the user.
+public let HomeDirectoryCharacter: Character = "~"
+
+
+/// Represents an absolute file system path, independently of what (or whether
+/// anything at all) exists at that path in the file system at any given time.
+/// An absolute path always starts with a `/` character, and holds a normalized
+/// string representation.  This normalization is strictly syntactic, and does
+/// not access the file system in any way.
+public struct AbsolutePath {
+    /// Private implementation details, shared with the RelativePath struct.
+    private let _impl: PathImpl
+    
+    /// Initializes the AbsolutePath from `string`, which must be an absolute
+    /// path (which means that it must begin with either a path separator or a
+    /// tilde, denoting an absolute path of a user account home directory).
+    ///
+    /// The path string will be normalized by:
+    /// - Collapsing `..` path components
+    /// - Removing `.` path components
+    /// - Expanding `~` or `~user` prefix
+    /// - Removing any trailing path separator
+    /// - Removing any redundant path separators
+    ///
+    /// This string manipulation may change the meaning of a path if any of the
+    /// path components are symbolic links on disk.  However, the file system is
+    /// never accessed in any way when initializing an AbsolutePath.
+    public init(_ string: String) {
+        precondition(string.characters.first == PathSeparatorCharacter
+                  || string.characters.first == HomeDirectoryCharacter)
+        
+        // Get a hold of the character view.
+        // FIXME: We need to investigate what is most performant here.
+        //        For example, is UTF-8 view more efficient, or unichars, etc?
+        var chars = string.characters
+        
+        // Expand any `~` or `~user` prefix.
+        if chars.first == HomeDirectoryCharacter {
```

Comment:
I see your point in theory, but that means that every client needs to remember to do this resolution individually on any path that can come in from the user.  That could lead to easily lead to cases where we end up with unexpanded paths starting with `~` characters.

Perhaps there is a pragmatic distinction to be made between ~ and ~user, since the former is not likely to require network access (even if the user's home directory is on the network, the actual characters that comprise the path are available locally; ~user is of course different, and perhaps that should be disallowed).